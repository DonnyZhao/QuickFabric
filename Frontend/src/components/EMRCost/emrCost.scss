.emr-cost-component {
  padding-top: 20px;
  margin-left: 90px;

  .button-list {
    padding: 10px 0px;
  }

  .pt-tab-list {
    margin-top: 20px;
  }

  .line {
    border-bottom: solid 1px #CED9E0;
  }

  .value {
    margin-top: -4px;
    margin-left: 20px;
  }

  .cell {
    width: 10px;
    height: 10px;
    border-radius: 50%;
    border: 1px solid;
    float: left;
  }
  
  .positive {
    background-color: #53B700;
    border-color: #53B700;
  }

  .negative {
    background-color: #DB3737;
    border-color: #DB3737;
  }

}

.pt-tab:focus {
  outline: 0;
}

.tab-button {
  background-color: none;
}

.exportBtn {
  width: 160px;
  height: 40px;
  font-size: 14px;
  background-color: white;
  border: solid #0097E6;
  border-width: 2px;
  border-radius: 20px;
  font-weight: 500;
  color: #0097E6;

  span {
    position: absolute;
    margin-top: 7px;
    margin-left: -47px;
  }
}

.material-icons {
  float: left
}
