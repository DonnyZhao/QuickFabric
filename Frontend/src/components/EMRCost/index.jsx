import React, { Component } from 'react'
import { Grid, GridColumn as Column } from '@progress/kendo-react-grid'
import { Spinner } from '@blueprintjs/core'
import emrHealthCell from '../EMRHealth/emrHealthCell'
import './emrCost.scss'
import '../EMRHealth/emrHealth1.scss'

import { withState } from '../../utils/components/with-state.jsx';
import { LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip } from 'recharts';

const StatefulGrid = withState(Grid);

/**
 * EMR Cost component: Table containing cost data with a final column of cost trend chart visualization.
 */
export default class EMRCost extends Component{
    account_range = '';
    constructor(props){
        super(props)
        
        this.state = {
            data: this.props.data,
            show: false,
            primaryTab: 'all'
        }

        this.renderTableData = this.renderTableData.bind(this)
        this.getTableData = this.getTableData.bind(this);
    }

    render = () => {
        return (
        <div className="emr-cost-component">
            <h2>EMR <i className="material-icon title-arrow">double_arrow</i> <span style={{ color: "#0097E4" }}>Cost</span></h2> 
            {
              this.props.fetching || this.props.data.length === 0? 
                <Spinner /> :
                this.renderTableData(this.props.data)
            }
        </div>
        )
    }

    renderTableData = data => {
      // Modifying data: Adding costs per month array as columns for table to consume.
      let modifiedData = this.getTableData(data);
      return (
        <StatefulGrid
          data={modifiedData.newData}
          style={{ "marginTop": 30 , height: "90vh", fontSize: '13px' }}
        >
          <Column field="emrGroup" title="EMR Group" width="150px" />
          <Column field="account" title="Account" width="120px" />
          <Column field="segment" title="Segment" width="100px" />
          <Column field="businessOwner" title="Business Owner" width="150px" />
          {modifiedData.costTitles.map((title, index) => {
            return (
            <Column field={title} title={title} width="80px" cell={(props) =>
              props.dataItem[title] == "0" ?
              <td>
                <span>N/A</span>
              </td> :
              <td>
                <span>${props.dataItem[title]}</span>
              </td>
            } />
            )
          })}
          <Column title="Cost Trend" width="auto" 
            cell={(props) => 
              <td>
                <LineChart width={480} height={120} data={props.dataItem.costPerMonth.reverse()}
                    >
                    <XAxis dataKey="billMonth" />
                    <YAxis />
                    <CartesianGrid strokeDasharray="3 3"/>
                    <Tooltip />
                    <Line type="monotone" dataKey="cost" stroke="#0097E4" activeDot={{r: 8}}/>
                </LineChart>
              </td>
            } 
          />
        </StatefulGrid>
      )
    }

  getTableData = data => {
    data = this.shortlistData(data.emrGroupCost);
    let newData = []
    let costTitles = []
    data.map((row, key) => {
      let vals = {
        emrGroup: row.emrGroup,
        account: row.account,
        segment: row.segment,
        businessOwner: row.businessOwner,
        costPerMonth: row.costPerMonth
      }
      row.costPerMonth.map((c) => {
        vals[c.billMonth] = c.cost;
        if (!costTitles.includes(c.billMonth))
          costTitles.push(c.billMonth)
      })
      newData.push(vals)
    })
    return {
      newData: newData,
      costTitles: costTitles
    }
  }
  
  shortlistData = data => {
    if (this.props.superAdmin) return data;
    let emrRoles = this.props.dccRoles.filter(x => x.serviceType === 'EMR')[0];

    if (emrRoles === undefined) return [];
  
    let roleSegments = [];
    emrRoles.roles.forEach((value) => {
        value.segments.forEach(x => {
          roleSegments.push(x);
        })
    });
    let roleMap = {};
    roleSegments.forEach((value, index) => {
      let accounts = [];
      value.accounts.forEach((act, index) => {
        accounts.push(act.accountId);
      })
      roleMap[value['segmentName'].toLowerCase()] = accounts;
    })
    
    let newData = data;
    newData = newData.filter((value) => {
      let type = value.segment;
      if (type in roleMap) {
        if (roleMap[type].includes(value.account)) {
          return value;
        }
      }
    });
  
    return newData;
  }
}