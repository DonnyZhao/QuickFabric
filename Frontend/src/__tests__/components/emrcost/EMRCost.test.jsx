import EMRCost from 'Components/EMRCost';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

const data = {
                emrGroupCost: [
                    {
                        emrGroup: "exploratory-sales",
                        account: "100000000000",
                        segment: "sales",
                        businessOwner: "The Boss",
                        costPerMonth: [
                            {
                                billMonth: "2019-09",
                                cost: 0
                            }, {
                                billMonth: "2019-10",
                                cost: 0
                            }
                        ]
                    }
                ]
            };

const uiListArray = {
    roles: [],
    accounts: [],
    response: {
        segments: []
    }
}

const component = shallow(<EMRCost  
                            uiListData={uiListArray}
                            superAdmin={true}
                            data={data}
                            />)

describe('Testing EMR Cost Table', () => {
    it('Should render the component', () => {
        expect(toJson(component)).toMatchSnapshot();
    })

    it('Should render the table with data', () => {
        const rows = component.find('StatefulGrid');
        expect(rows.length).toBe(1);
    })
})