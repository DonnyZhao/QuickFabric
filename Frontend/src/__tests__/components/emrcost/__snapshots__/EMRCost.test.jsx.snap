// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing EMR Cost Table Should render the component 1`] = `
<div
  className="emr-cost-component"
>
  <h2>
    EMR 
    <i
      className="material-icon title-arrow"
    >
      double_arrow
    </i>
     
    <span
      style={
        Object {
          "color": "#0097E4",
        }
      }
    >
      Cost
    </span>
  </h2>
  <StatefulGrid
    data={
      Array [
        Object {
          "2019-09": 0,
          "2019-10": 0,
          "account": "100000000000",
          "businessOwner": "The Boss",
          "costPerMonth": Array [
            Object {
              "billMonth": "2019-09",
              "cost": 0,
            },
            Object {
              "billMonth": "2019-10",
              "cost": 0,
            },
          ],
          "emrGroup": "exploratory-sales",
          "segment": "sales",
        },
      ]
    }
    style={
      Object {
        "fontSize": "13px",
        "height": "90vh",
        "marginTop": 30,
      }
    }
  >
    <Component
      editable={true}
      field="emrGroup"
      filterable={true}
      groupable={true}
      reorderable={true}
      resizable={true}
      sortable={true}
      title="EMR Group"
      width="150px"
    />
    <Component
      editable={true}
      field="account"
      filterable={true}
      groupable={true}
      reorderable={true}
      resizable={true}
      sortable={true}
      title="Account"
      width="120px"
    />
    <Component
      editable={true}
      field="segment"
      filterable={true}
      groupable={true}
      reorderable={true}
      resizable={true}
      sortable={true}
      title="Segment"
      width="100px"
    />
    <Component
      editable={true}
      field="businessOwner"
      filterable={true}
      groupable={true}
      reorderable={true}
      resizable={true}
      sortable={true}
      title="Business Owner"
      width="150px"
    />
    <Component
      cell={[Function]}
      editable={true}
      field="2019-09"
      filterable={true}
      groupable={true}
      reorderable={true}
      resizable={true}
      sortable={true}
      title="2019-09"
      width="80px"
    />
    <Component
      cell={[Function]}
      editable={true}
      field="2019-10"
      filterable={true}
      groupable={true}
      reorderable={true}
      resizable={true}
      sortable={true}
      title="2019-10"
      width="80px"
    />
    <Component
      cell={[Function]}
      editable={true}
      filterable={true}
      groupable={true}
      reorderable={true}
      resizable={true}
      sortable={true}
      title="Cost Trend"
      width="auto"
    />
  </StatefulGrid>
</div>
`;
